cpp/src/arrow/engine/substrait/relation_internal.cc

```diff
@@ -162,36 +170,40 @@ Result<DeclarationInfo> FromProto(const substrait::Rel& rel, const ExtensionSet&
         }
 
         path = path.substr(7);
-        if (item.path_type_case() ==
-            substrait::ReadRel_LocalFiles_FileOrFiles::kUriPath) {
-          ARROW_ASSIGN_OR_RAISE(auto file, filesystem->GetFileInfo(path));
-          if (file.type() == fs::FileType::File) {
-            files.push_back(std::move(file));
-          } else if (file.type() == fs::FileType::Directory) {
+        switch (item.path_type_case()) {
+          case substrait::ReadRel_LocalFiles_FileOrFiles::kUriPath: {
```

Comment:
I kind of prefer `substrait::ReadRel::LocalFiles::FileOrFiles` over `substrait::ReadRel_LocalFiles_FileOrFiles` but I see we have some precedence for this case so I don't think you have to fix it for this PR.

Reply:
Good note. I also saw this kind of style in a few places. May be I will make a cleanup PR after to make them uniform across Substrait. 

JIRA: https://issues.apache.org/jira/browse/ARROW-17647